core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -245,13 +250,13 @@ private PartitionSpec spec(int specId) {
 
   /** Add a delete file to the new snapshot. */
   protected void add(DeleteFile file) {
-    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    validateNewDeleteFile(file);
     add(new PendingDeleteFile(file));
   }
 
   /** Add a delete file to the new snapshot. */
   protected void add(DeleteFile file, long dataSequenceNumber) {
-    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    validateNewDeleteFile(file);
```

Comment:
I'm not quite clear if there's a case if a manifest rewrite could cause an issue here.  Is it possible that just rewriting delete manifests as part of incremental maintenance would hit the table version check?  I don't feel like we can require delete file rewrite as part of that.

Reply:
This only affects new delete files that are added to the table. I had a test in the prototype PR that verified that automatic manifest compaction after upgrading to V3 didn't have issues. Let me check if I added it here too.

Reply:
Yep, it is [here](https://github.com/apache/iceberg/pull/11495/files#diff-410ff1b47d9a44a2fd5dbd103cad9463d82c8f4f51aa1be63b8b403123ab6e0eR1571). Let me know if that's what you meant.

Reply:
That said, there are some test failures in Spark that may be related. Let me explore.

Reply:
It was another issue.

Reply:
Thanks @aokolnychyi for the background. It wasn't clear to me the purpose of `validateNewDeleteFile()`during `add`.

Reply:
I added more testing around manifest rewrites is Spark.